rfcs/20210504-kernel-extension-variable-ops.md

```diff
@@ -0,0 +1,218 @@
+// TF_MaybeLockVariableInputMutexesInOrder API
+TF_CAPI_EXPORT extern void TF_ReleaseVariableInputLockHolder(
+                                  TF_VariableInputLockHolder* lockHolder);
+
+
+```
+
+*Resource Variables*
+ 
+Below mentioned APIs expose functionality in Core TensorFlow to allow plugins
+to do Assign and Update operations to implement different ResourceVariable
+primitives. These are higher level operations which can be easily integrated
+into the plugin by directly calling these APIs from the Compute function of
+these ops in the Plugin. 
+ 
+```
+// Expose higher level Assignment operation for Pluggable vendors to implement
+// in the plugin for Training. The API takes in the context with indices for
+// the input and value tensors. It also accepts the copy functor provided by
+// pluggable vendor to do the copying of the tensors.
+TF_CAPI_EXPORT extern void TF_AssignVariable(TF_OpKernelContext* ctx,
+                               int input_index,
+                               int value_index,
+                               void (*copyFunc)(TF_OpKernelContext * ctx,
+                                                TF_Tensor *source,
+                                                TF_Tensor *dest),
+                               TF_Status* status);
+
+// Expose higher level AssignUpdate operation for Pluggable vendors to implement
+// in the plugin for Training. The API takes in the context with indices for
+// the input and value tensors. It also accepts the copy/update functor provided by
+// pluggable vendor to perform these operations respectively.
+TF_CAPI_EXPORT extern void TF_AssignUpdateVariable(
+                                    TF_OpKernelContext* ctx,
+                                    int input_index,
+                                    int value_index,
+                                    int Op,
+                                    int isVariantType,
+                                    void (*copyFunc)(TF_OpKernelContext * ctx, 
+                                                     TF_Tensor *source, 
+                                                     TF_Tensor *dest),
+                                    void (*updateFunc)(TF_OpKernelContext *ctx,
+                                                       TF_Tensor *tensor,
+                                                       TF_Tensor *value, int Op),
+                                    TF_Status* status);
+
+```
+
+### Alternatives Considered
+
+We considered two different ways to add  support for resource variables and
+optimizers operations. Option #1 is to expose the required lower level
+structures in Tensorflow core like TF_Mutex, TF_ResourceHandle, TF_RefCountPtr
+to the plugin vendors. This will allow plugin writers the flexibility to
+construct higher level optimizer operations using these lower level primitives
+and will be scalable for newer operations. Option #2, is to expose all the
```

Comment:
Do you have any near-term plans to expose any other ops? If so option #1 maybe worth paying more heed to. You already need to expose `TF_GetInputTensorFromVariable` for the training ops which does most of what you need functionality-wise. Although, I agree the current API is much cleaner.

Reply:
Passing on comments from @Jianhui-Li and @jzhoulon: They prefer option #1 for scalability. (Please feel free to elaborate here.)

Reply:
@saxenasaurabh , for resource variable ops:

1. some of them fall in the DEVICE_DEFAULT category: like `ReadVariableOp`, [`VariableShapeOp`](https://github.com/tensorflow/tensorflow/pull/49111/files#diff-c7a7ce394ac5cab8d4e036b91194fedd367ab1366283740479cddd165d184235R330), `DestroyResourceOp`, 
2. ResourceGather , ResourceScatter as discussed above we have implemented using the `GetInputTensorFromVariable` and `LockMutexesInOrder`
3. @saxenasaurabh , that's a good point. AssignUpdateVariable and AssignVariable are currently implemented as an Op, we can use `TF_GetInputTensorFromVariable` and `TF_GetTrainingVariableMutex` (this is a helper function we use for `TF_MaybeLockVariableInputMutexesInOrder`) to implement this functionality. Depending on whether the Op needs exclusive lock we can use the `lock` or `shared_lock` to implement it. This is part of the helper function `TF_MaybeLockVariableInputMutexesInOrder`. 

This covers majority of the resource variable operations we have encountered in the Training networks.

Reply:
@Jianhui-Li and @jzhoulon thanks for your feedback. I agree Option 1 is more scalable. We started at the higher level interface for few reasons:

1. Most of the common resource variable ops and training ops exhibited a pattern to implement them which we were mostly able to capture using these APIs.
2. We wanted to keep a smaller / tight interface to keep the code maintainable and allows us to organically grow in the future if we see the need.
3. With the upcoming new stack of MLIR/TFRT (As mentioned in the [StreamExecutor RFC](https://github.com/tensorflow/community/blob/master/rfcs/20200612-stream-executor-c-api.md)), we wanted to keep the design of interface simple and high-level to represent the needed ops.

Reply:
@kulinseth, @jzhoulon and I discussed and agreed that Option 2 is simpler but we need to cover the other cases we discovered (TensorArray and UnBatchResource), which use resource manager and are dispatched to device.  Covering these two cases with extra APIs following Option 2 approach is more manageable than we originally thought, so we don't really against Option 2.   Could you please also verify whether the Option 2 approach can cover these cases?

Reply:
@Jianhui-Li thanks for getting back!. That's a good point, initially we weren't looking into `TensorArray` and `UnbatchResource`, as we didn't encounter them in performance critical sections of the networks we were looking at. I haven't dug into the details, from cursory glance it seems to me we might need to expose `TF_TensorArray` and some methods handling `TF_TensorArray` type. 

Similarly for `UnbatchResource` as you mentioned we would need to expose `ResourceManager` details. Even then I am not sure it will cover all the necessary functionality due to some of the dependency on `AsyncOpKernel` and `AdaptiveBatchScheduler`. I feel these are bigger API changes which might need an RFC of their own. 

@saxenasaurabh , @Jianhui-Li , @penpornk , @jzhoulon Not sure, what do you guys think?